dlt/extract/reference.py

```diff
@@ -0,0 +1,229 @@
+from abc import ABC, abstractmethod
+from importlib import import_module
+from typing import (
+    Callable,
+    Dict,
+    List,
+    Any,
+    Generic,
+    Tuple,
+    cast,
+    overload,
+)
+from typing_extensions import Self, TypeVar
+from typing import Dict, Type, ClassVar
+
+from dlt.common import logger
+from dlt.common.configuration.specs import BaseConfiguration, known_sections
+
+from dlt.common.runtime.run_context import get_plugin_modules
+from dlt.common.schema import Schema
+from dlt.common.schema.typing import TSchemaContract
+from dlt.common.typing import ParamSpec
+from dlt.common.exceptions import MissingDependencyException
+from dlt.common.warnings import Dlt04DeprecationWarning, deprecated
+
+from dlt.extract.source import DltSource
+from dlt.extract.exceptions import UnknownSourceReference
+
+TDltSourceImpl = TypeVar("TDltSourceImpl", bound=DltSource, default=DltSource)
+TSourceFunParams = ParamSpec("TSourceFunParams")
+
+
+class SourceFactory(ABC, Generic[TSourceFunParams, TDltSourceImpl]):
+    def __call__(
+        self, *args: TSourceFunParams.args, **kwargs: TSourceFunParams.kwargs
+    ) -> TDltSourceImpl:
+        """Makes dlt source"""
+        pass
+
+    # TODO: make factory to expose SourceReference with actual spec, name and section
+    # model after Destination, which also needs to be broken down into reference and factory
+    ref: "SourceReference"
+
+    @abstractmethod
+    def clone(
+        self,
+        *,
+        name: str = None,
+        section: str = None,
+        max_table_nesting: int = None,
+        root_key: bool = False,
+        schema: Schema = None,
+        schema_contract: TSchemaContract = None,
+        spec: Type[BaseConfiguration] = None,
+        parallelized: bool = None,
+        _impl_cls: Type[TDltSourceImpl] = DltSource,  # type: ignore[assignment]
+    ) -> Self:
+        """Overrides default decorator arguments that will be used to when DltSource instance and returns modified clone."""
+
+    with_args = deprecated("Please use clone method instead", category=Dlt04DeprecationWarning)(
+        clone
+    )
+
+
+AnySourceFactory = SourceFactory[Any, DltSource]
+
+
+class SourceReference:
```

Comment:
Maybe these classes are to different, but isn't there a common mechanism between the source and the destination reference that would warrant a mutual baseclass?